lib/did_you_mean/spell_checker.rb

```diff
@@ -10,23 +10,24 @@ def initialize(dictionary: )
     end
 
     def correct(input)
-      input     = normalize(input)
-      threshold = input.length > 3 ? 0.834 : 0.77
+      input        = normalize(input)
+      input_length = input.length
+      threshold    = input_length > 3 ? 0.834 : 0.77
 
-      words = @dictionary.select {|word| JaroWinkler.distance(normalize(word), input) >= threshold }
-      words.reject! {|word| input == word.to_s }
+      words = @dictionary.reject {|word| input == word.to_s }
+      words.select! {|word| JaroWinkler.distance(normalize(word), input) >= threshold }
```

Comment:
It's important to call `#select` first because that's how the spellchecker deals with outliers. This would change the order of calculation from `O(n)` and `O(m)` to `2 * O(n)` where `n` is the dictionary size (generally between 100 and 300) and `m` is the number of possible suggestions (generally below 20). The memory usage of the spell checker also went up from [769KB](https://travis-ci.org/yuki24/did_you_mean/jobs/324276683#L718) to about [1.5MB](https://travis-ci.org/yuki24/did_you_mean/jobs/327300205#L712) (script can be found [here](https://github.com/yuki24/did_you_mean/blob/master/benchmark/memory_usage.rb)).